bluemira/equilibria/optimisation/problem/base.py

```diff
@@ -189,26 +199,58 @@ def update_magnetic_constraints(
                 constraint._args["scale"] = self.scale
 
     def _make_numerical_constraints(
-        self,
+        self, coilset: CoilSet
     ) -> tuple[list[ConstraintT], list[ConstraintT]]:
         """Build the numerical equality and inequality constraint dictionaries."""
         if (constraints := getattr(self, "_constraints", None)) is None:
             return [], []
         equality = []
         inequality = []
         for constraint in constraints:
-            f_constraint = constraint.f_constraint()
+            f = constraint.f_constraint()
+
+            f_c = f.f_constraint
+            df_c = getattr(f, "df_constraint", None)
+
+            if coilset._contains_circuits:
+                # if the coilset contains circuits, we need to wrap the constraint
+                # functions (f_c) and 'expand' the current vector, which repeats
+                # the currents for each circuit in the coilset.
+                #
+                # for the derivative function (df_c), we also apply the "expand" matrix
+                # to the output of the derivative function which reduces the
+                # shape of the output to the shape of the coilset opt currents
+                # by adding (summing) the derivatives from the coils in the
+                # circuit.
+
+                # wrap the constraint function
+                @functools.wraps(f.f_constraint)
+                def wrapped_f_c(x, f=f):
+                    return f.f_constraint(coilset._opt_currents_expand_mat @ x)
+
+                f_c = wrapped_f_c
+
+                if df_c is not None:
+                    # wrap the derivative function
+                    @functools.wraps(f.df_constraint)
+                    def wrapped_df_c(x, f=f):
+                        df_res = f.df_constraint(coilset._opt_currents_expand_mat @ x)
+                        return df_res @ coilset._opt_currents_expand_mat
+
+                    df_c = wrapped_df_c
+
             d: ConstraintT = {
                 "name": f_constraint.name,
-                "f_constraint": f_constraint.f_constraint,
-                "df_constraint": getattr(f_constraint, "df_constraint", None),
+                "f_constraint": f_c,
+                "df_constraint": df_c,
                 "tolerance": constraint.tolerance,
             }
             # TODO: tidy this up, so the interface guarantees this works!
```

Comment:
Just wondering what this #TODO is from? Do we still need it?

Reply:
Yeah.. It's because the ABC doesn't have a good interface. It should all be improved.